components/driver/gpio.c

```diff
@@ -88,6 +88,11 @@ esp_err_t gpio_set_intr_type(gpio_num_t gpio_num, gpio_int_type_t intr_type)
     GPIO_CHECK(GPIO_IS_VALID_GPIO(gpio_num), "GPIO number error", ESP_ERR_INVALID_ARG);
     GPIO_CHECK(intr_type < GPIO_INTR_MAX, "GPIO interrupt type error", ESP_ERR_INVALID_ARG);
     GPIO.pin[gpio_num].int_type = intr_type;
+    if (intr_type == GPIO_INTR_DISABLE) {
+        gpio_intr_disable(gpio_num);
+    } else {
+        gpio_intr_enable(gpio_num);
```

Comment:
I don't think we want to automatically enable the interrupt when setting the type.

Reply:
@negativekelvin Thanks for your review. It's almost been 7 months, since I had my head into this and I unfortunately do not have time to help out on this project at the moment. I however remembered that what you spotted here, bothered me as well. It was however the solution that I found with the bare minimum changes to the "structure", at that time. I wanted to rewrite a lot more, but I have had no response on my patch, from anyone at all, so it now is what it is.

BTW, just 2 days ago, @koobest wrote the following [here](https://github.com/espressif/esp-idf/issues/2845#issuecomment-513447994):

> We are use the GPIO interrupt raw register in the ISR. So, When the interrupt is disabled, the interrupt raw status register will still be set when the interrupt trigger signal arrives.By the way, when the int_type register is set to 0, the interrupt raw status register will not be set. It is recommended to use gpio_set_intr_type to enable or disable interrupts.
thanks !!

I'm not sure what to make of his last sentence, but if @koobest is right, then my patch should probably stay as it is.

I would suggest to rework the whole interrupt enabling and disabling thing, and name things according to what they do, once #2845 is solved and stable.

Reply:
If you look at the follow-up to that you'll see it will be changing but I see your point about forcing the enable since the current code will read the flag anyway.  It would be the least bad option if there wasnt a fix.

Reply:
Hi, @wingunder 
Thanks for your contribution.  Can you help post this PR based on the latest master branch? so we can help review and merge into the master branch.

> We are use the GPIO interrupt raw register in the ISR. So, When the interrupt is disabled, the interrupt raw status register will still be set when the interrupt trigger signal arrives.By the way, when the int_type register is set to 0, the interrupt raw status register will not be set. It is recommended to use gpio_set_intr_type to enable or disable interrupts.
thanks !!

Here I mentioned a bug in GPIO driver.  If we want to disable a GPIO's interrupt , we can use `gpio_set_intr_type(gpio_num, GPIO_INTR_DISABLE)`, and do not use `gpio_intr_disable(gpio_num)`.  But  this is just a work around way, the bug still exists.

Now, we want to fix it. we should polling the interrupt status register instead of the raw status register.

Reply:
@koobest At the moment I really cannot help you out, due to my heavy work load and private life. I'm sorry, but if I am really optimistic, I could eventually look at it by the end of September 2019. This is however _not_ a promise.

I'm sure that there are some other capable contributors around that could help you out.

Reply:
Hi, @wingunder 
It doesn't matter :-),  I just want to try to merge your PR into the master, because you spent your time on it. we can fix this bug by ourselves. 

Thank you.

Reply:
>I just want to try to merge your PR into the master

@koobest, No problem, just 'git format-patch' and 'git am' my commits onto your bugfix branch and continue. It shouldn't be too difficult to fix the conflicts.